File: 0234_Palindrome_Linked_List/solution.go
/*
234. Palindrome Linked List

Level: Easy

https://leetcode.com/problems/palindrome-linked-list
*/
package leetcode0234

import (
	. "../common"
)

/*

 */
 func isPalindrome(head *ListNode) bool {
	nums := make([]int, 0, 64)
	for head != nil {
		nums = append(nums, head.Val)
		head = head.Next
	}

	l, r := 0, len(nums)-1
	for l < r {
		if nums[l] != nums[r] {
			return false
		}
		l++
		r--
	}
	return true
}
/*
time O(n), space O(1)
*/
func isPalindromeBetter(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	slow, fast := head, head.Next
	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
	}

	secondHead := slow.Next
	secondHead = reverseList(secondHead)

	for head != nil && secondHead !=nil {
		if head.Val != secondHead.Val {
			return false
		}
		head = head.Next
		secondHead = secondHead.Next
	}

	return true
}

func reverseList(head *ListNode)*ListNode{
	if head == nil || head.Next == nil {
		return head
	}

	var prev *ListNode
	for head != nil {
		next:= head.Next
		head.Next = prev
		prev = head
		head = next
	}

	return prev
}

File: 0234_Palindrome_Linked_List/solution.cpp
/*
234. Palindrome Linked List

Level: Easy

https://leetcode.com/problems/palindrome-linked-list


- 206. Reverse Linked List

*/
#include <vector>
#include <string>
#include <unordered_set>

#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  bool isPalindrome(ListNode *head)
  {
    if (head == nullptr || head->next == nullptr)
      return true;

    //find list center
    ListNode *fast = head;
    ListNode *slow = head;

    while (fast->next != nullptr && fast->next->next != nullptr)
    {
      fast = fast->next->next;
      slow = slow->next;
    }

    ListNode *secondHead = slow->next;
    slow->next = nullptr;

    //reverse second part of the list
    ListNode *prev = nullptr;
    ListNode *curr = secondHead;

    while (curr)
    {
      ListNode *next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }

    //compare two sublists
    ListNode* p = prev;
    ListNode* q = head;
    while (p)
    {
      if (p->val != q->val)
        return false;

      p = p->next;
      q = q->next;
    }

    return true;
  }
};